compiler/parser/RdrHsSyn.hs

```diff
@@ -3166,64 +3174,141 @@ validateTemplate vtTemplateName tbd@TemplateBodyDecls{..}
-    validateImplementsBlock (L loc (ParsedImplementsDeclBlock iface decls)) = do
-      impls <- mapM validateMethodImpl (groupValDecls decls)
-      checkMultipleDecls iface impls
-      mapM_ (checkNumArgs iface) impls
+-- | Convert controlled choice groups to a list of individual choices with controllers.
+-- Leave type variable information empty, to be added afterwards.
+choiceGroupsToCombinedChoices
+  :: [Located (LHsExpr GhcPs, Located [Located ChoiceData])]
+  -> [CombinedChoiceData]
+choiceGroupsToCombinedChoices = concatMap distributeController
+  where
+    distributeController (L _ (controller, L _ choices)) = map (makeCombinedChoice controller) choices
+    makeCombinedChoice controller choice = CombinedChoiceData controller Nothing (unLoc choice) False TemplateChoice
 
-      let isViewImpl impl = rdrNameToString (vimdId impl) == "view"
-      let (viewImpls, nonViewImpls) = partition (isViewImpl . unLoc) impls
-      viewImpl <- case viewImpls of
-        [] -> pure Nothing -- TODO: Fail on missing viewtype when PR #14486 merged in & tests fixed
-        [viewImpl] -> pure $ Just viewImpl
-        _ -> addFatalError loc (text "implements block has more than one view method")
+data TemplateOrInterface a b
+  = TorI_Template a
+  | TorI_Interface b
+
+type TemplateOrInterface' a = TemplateOrInterface a a
+
+validateInterfaceInstances ::
+       TemplateOrInterface' (Located RdrName)
+  ->   [Located ParsedInterfaceInstance]
+  -> P [Located ValidInterfaceInstance]
+validateInterfaceInstances parent = mapM (validateInterfaceInstance parent)
+
+validateInterfaceInstance ::
+       TemplateOrInterface' (Located RdrName)
+  ->   Located ParsedInterfaceInstance
+  -> P (Located ValidInterfaceInstance)
+validateInterfaceInstance parent (L loc piib) = do
+  checkParent
+  impls <- mapM validateMethodImpl (groupValDecls (unLoc piiDefs))
+  checkMultipleDecls impls
+  mapM_ checkNumArgs impls
+
+  let (viewImpls, nonViewImpls) = partition isViewImpl impls
+
+  viewImpl <- checkViewImpl viewImpls
+
+  pure $ L loc ValidInterfaceInstance
+    { viiInterface = piiInterface
+    , viiTemplate = piiTemplate
+    , viiDefs = nonViewImpls
+    , viiView = viewImpl
+    }
+  where
+    ParsedInterfaceInstance
+      { piiInterface
+      , piiTemplate
+      , piiDefs
+      } = piib
+
+    iiFatalError :: SrcSpan -> SDoc -> P a
+    iiFatalError loc err =
+      addFatalError loc $ vcat
+        [ err
+        , text "in"
+          <+> quotes
+            ( text "interface instance"
+              <+> ppr piiInterface
+              <+> text "for"
+              <+> ppr piiTemplate
+            )
+        , text "in" <+> pprParent
+        ]
+
+    pprParent = case parent of
+      TorI_Template t -> text "template" <+> quotes (ppr t)
+      TorI_Interface i -> text "interface" <+> quotes (ppr i)
+
+    -- NOTE(MA): this only really checks the 'RdrName's are identical,
+    -- so it doesn't look through qualified types or type synonyms.
```

Comment:
Do we need to worry about qualified types? The interface or template is defined in the current module so it seems like that should always be unqualified anyway. Type synonyms are a fair point but it seems pretty weird to want to use a synonym for that.

Reply:
That was my thought process too, it would be weird to qualify/alias the name of the parent in a declaration _in the parent_. 

The one use case I could think of would be an `interface instance` in the definition of template/interface `Foo` in a module that imports another `Foo`. However that fails with `Ambiguous occurrence ‘Foo’` errors even without the `interface instance`, due to the `HasX` instances for templates.